File: telegram-bot-core/src/main/java/ru/tggc/telegrambotcore/registry/resolver/HandlerArgumentResolver.java
package ru.tggc.telegrambotcore.registry.resolver;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.Arrays;
import java.util.List;

@Component
@RequiredArgsConstructor
public class HandlerArgumentResolver {
    private final List<ParameterResolver<?>> parameterResolvers;

    public Object[] resolve(Method method, HandlerCtx ctx) {
        return Arrays.stream(method.getParameters())
                .map(p -> resolveParameter(p, ctx))
                .toArray();
    }

    private Object resolveParameter(Parameter parameter, HandlerCtx ctx) {
        return parameterResolvers.stream()
                .filter(r -> r.supports(parameter))
                .findFirst()
                .map(r -> r.resolve(parameter, ctx))
                .orElse(null);
    }
}

File: telegram-bot-core/src/main/kotlin/ru/tggc/telegrambotcore/registry/resolver/HandlerArgumentResolver.kt
package ru.tggc.telegrambotcore.registry.resolver

import org.springframework.stereotype.Component
import java.lang.reflect.Method
import java.lang.reflect.Parameter

@Component
class HandlerArgumentResolver(private val parameterResolvers: MutableList<ParameterResolver<*>>) {

    fun resolve(method: Method, ctx: HandlerCtx): Array<Any?> =
        method.parameters
            .map { resolveParameter(it, ctx) }
            .toTypedArray()

    private fun resolveParameter(parameter: Parameter?, ctx: HandlerCtx): Any? =
        parameterResolvers
            .firstOrNull { it.supports(parameter) == true }
            ?.resolve(parameter, ctx)

}
